src/main/java/com/google/api/generator/engine/ast/CastExpr.java

```diff
@@ -52,17 +52,16 @@ public CastExpr build() {
       // Only numeric types can be casted around.
       if (castType.isPrimitiveType() && exprType.isPrimitiveType()) {
         // Check that the types are either both booleans or both not booleans.
-        // TODO(xiaozhenliu): This is one of the places where we need to check the equality of boxed
-        // and primitive types.
         Preconditions.checkState(
             !(castType.equals(TypeNode.BOOLEAN) ^ exprType.equals(TypeNode.BOOLEAN)),
             "Numeric and boolean types are not inter-castable");
       } else {
+        boolean isReferenceTypeEquals =
```

Comment:
This logic reads to me as "The cast type is a reference and the expression type is either null or a reference type." Could we please update the name to something that better reflects this? And/or break it down further so we have a variable for the expression type check only?

Also, what is the rationale for the current expression type check instead of `!isPrimitiveType && !equals(TypeNode.VOID)`?

Reply:
Broke the condition check further down to `exprType_referenceOrNull` and `validReferenceTypeCast`. ReferenceTypeCast is only valid when castType is referenceType AND exprType is NULL or reference.

`!TypeNode.isPrimitiveType(exprType) && !equals(TypeNode.VOID)` only ensures the TypeKind is `Object`, but misses the check for `exprType.reference()` is not null.

Reply:
Thanks, this is clearer. `isValidPrimitiveTypeCast` seems to imply that `void` should be disallowed, could we add a test to ensure this is the case?

Reply:
yes, `isValidPrimitiveTypeCast` ensures that void type is not allowed here.
 While we cannot create a variable with `void` type since we have condition check in Variable: `!variable.type().typeKind().equals(TypeNode.TypeKind.VOID), "Variable type cannot be void"`.

Reply:
It doesn't though, because this doesn't look like it prevents `int x; (void) x`.

Reply:
Actually we take `void` as a primitive type in `isPrimitiveType()` for TypeNode, so we won't disallow void here. But the assertion above prevents that:
```
      Preconditions.checkState(
          !castType.equals(TypeNode.VOID) && !exprType.equals(TypeNode.VOID),
          "Casting cannot be done on or to void types");
```